src/lib/utils/tree_hash.h

```diff
@@ -0,0 +1,269 @@
+                        concepts::tree_address<Address, LayerIdx, NodeIdx> && concepts::strong_span<NodeSS> &&
+                        requires(T func, NodeSS out, const Address& address) {
+                           { func(out, address) };
+                        };
+
+}  // namespace concepts
+
+/**
+ * @brief Treehash logic to build up a merkle hash tree.
+ *
+ * Computes the root of the merkle tree.
+ * Can also output an authentication path necessary for a hash based signature.
+ *
+ * Given the following tree:
+ *  Layer:
+ *     2       7R
+ *            /  \
+ *     1    3X    6A
+ *         / \    / \
+ *     0  1X  2A 4   5
+ *
+ * The treehash logic traverses the tree (Post-order traversal), i.e., the nodes are
+ * discovered in order 1,2,3,...,7. If we want to create a signature using leaf node 1,
+ * the authentication path is (Node 2, Node 6), since we need those to compute the
+ * root.
+ *
+ * @param out_root An output buffer to store the root node in (size: node_size ).
+ * @param out_auth_path Optional buffer to store the authentication path in (size: node_size * total_tree_height).
+ * @param leaf_idx The optional index of the leaf used to sign in the bottom tree layer beginning with index 0.
+ *                 nullopt if no node is signed, so we need no auth path.
+ * @param node_size The size of each node in the tree.
+ * @param total_tree_height The hight of the merkle tree to construct.
+ * @param idx_offset If we compute a subtree this marks the index of the leftmost leaf node in the bottom layer
+ * @param node_pair_hash The function to process two child nodes to compute their parent node.
+ * @param gen_leaf The logic to create a leaf node given the address in the tree. Probably this function
+ *                 creates a one-time/few-time-signature's public key which is hashed to be the leaf node.
+ * @param tree_address The address that is passed to gen_leaf or node_pair hash. This function will update the
+ *                     address accordings to the currently processed node. This object may contain further
+ *                     algorithm specific information, like the position of this merkle tree in a hypertree.
+ */
+template <concepts::contiguous_strong_type TreeNode,
+          concepts::strong_span AuthPathSS,
+          concepts::tree_node_index TreeNodeIndex,
+          concepts::tree_layer_index TreeLayerIndex,
+          typename Address>
+   requires concepts::tree_address<Address, TreeLayerIndex, TreeNodeIndex>
+inline void treehash(
+   StrongSpan<TreeNode> out_root,
+   std::optional<AuthPathSS> out_auth_path,
+   std::optional<TreeNodeIndex> leaf_idx,
+   size_t node_size,
+   TreeLayerIndex total_tree_height,
+   uint32_t idx_offset,
+   concepts::tree_hash_node_pair<TreeNodeIndex, TreeLayerIndex, Address, StrongSpan<TreeNode>> auto node_pair_hash,
+   concepts::tree_gen_leaf<TreeNodeIndex, TreeLayerIndex, Address, StrongSpan<TreeNode>> auto gen_leaf,
+   Address& tree_address) {
```

Comment:
# 😲 

Any chance we could wrap that into a reasonable class that removes some complexity from this interface? Off the top of my head, I don't know enough to give useful suggestions. Though, on first glance, that seems like a nightmare to use and/or understand.

Can we go through the usage of this in a pair-programming session at one point please?

Reply:
Yeah, I admit that this construct looked much better in my head. The goal of this template hell is that the calling party only calls this function with its own strong types while all types of `treehash` are deduced automatically without the need to pass any template parameters. But yeah, I would love to get your input on this in a pair-programming session!